### PR TITLE
fix: waiting for proposal in pre-commit phase

### PR DESCRIPTION
**Pull request by b00f**
## Description

This pull request addresses a situation observed in the Testnet where certain nodes with slow connections become absentees in the certificate. This occurs when they receive block announcements before entering the pre-commit phase, possibly due to late proposals or votes.

To address this issue, the following changes have been made:

- When there are sufficient pre-commit votes but no proposal, the system now waits to receive the proposal before proceeding.
- If a proposal is received after committing a block, it is now accepted.
- If timeout expires in pre-commit phase, we move to the change-proposer phase.

These scenarios are covered by tests, and  also been tested using [TLA+](https://lamport.azurewebsites.net/tla/tla.html) to ensure its correctness.

### CODE REVIEW

File: consensus/consensus.go [hunk 1]
```diff
@@ -183,15 +178,26 @@ func (cs *consensus) SetProposal(p *proposal.Proposal) {
...
-	if err := cs.state.ValidateBlock(p.Block()); err != nil {
-		cs.logger.Warn("invalid block", "proposal", p, "err", err)
-		return
+	if p.Height() == cs.state.LastBlockHeight() {
+		// A slow node might receive a proposal after committing the proposed block.
+		// In this case, we accept the proposal and allow nodes to continue.
+		// By doing so, we enable the validator to broadcast its votes and
+		//prevent it from being marked as absent in the block certificate.
+		cs.logger.Trace("block is committed for this height", "proposal", p, "block hash", cs.state.LastBlockHash())
+		if p.Block().Hash() != cs.state.LastBlockHash() {
```
Comment (themantre) [inline consensus/consensus.go:187]: add warning log here

File: consensus/precommit.go [hunk 2]
```diff
@@ -23,7 +23,9 @@ func (s *precommitState) decide() {
 	precommitQH := precommits.QuorumHash()
 	if precommitQH != nil {
 		s.logger.Debug("precommit has quorum", "precommitQH", precommitQH)
-		s.enterNewState(s.commitState)
+		if s.hasVoted {
```
Comment (themantre) [inline consensus/precommit.go:26]: query proposal here. (don't forget to add a log)

**Review by themantre — CHANGES_REQUESTED** [review]

File: consensus/consensus.go (hunk 1 above)
Comment (b00f) [reply consensus/consensus.go:187]: done

File: consensus/precommit.go (hunk 2 above)
Comment (b00f) [reply consensus/precommit.go:26]: It already called `vote()` before and should be queried

File: consensus/consensus.go [hunk 3]
```diff
@@ -183,15 +178,27 @@ func (cs *consensus) SetProposal(p *proposal.Proposal) {
...
-	}
-
-	if err := cs.state.ValidateBlock(p.Block()); err != nil {
-		cs.logger.Warn("invalid block", "proposal", p, "err", err)
-		return
+	if p.Height() == cs.state.LastBlockHeight() {
+		// A slow node might receive a proposal after committing the proposed block.
+		// In this case, we accept the proposal and allow nodes to continue.
+		// By doing so, we enable the validator to broadcast its votes and
+		//prevent it from being marked as absent in the block certificate.
```
Comment (themantre) [inline consensus/consensus.go:185]:
```suggestion
		// prevent it from being marked as absent in the block certificate.
```

**Review by themantre — APPROVED** [review]

**Pull request merged**